File: backend/pkg/api/organization/handlers.go
package organization

import (
	"backend/pkg/api/middleware"
	"backend/pkg/api/response"
	"backend/pkg/api/validation"
	"backend/pkg/core"
	coreOrg "backend/pkg/core/organization"

	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"
)

// ListOrganizations handles GET /organizations
// @Summary List organizations
// @Description Get a list of all organizations
// @Tags organizations
// @Accept json
// @Produce json
// @Success 200 {object} response.SuccessResponse{data=[]dto.OrganizationResponse}
// @Failure 500 {object} response.SuccessResponse
// @Router /organizations [get]
func ListOrganizations(c *fiber.Ctx) error {
	orgs, err := coreOrg.List(c.Context())
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, orgs)
}

// GetOrganization handles GET /organizations/:id
// @Summary Get organization
// @Description Get an organization by ID
// @Tags organizations
// @Accept json
// @Produce json
// @Param id path string true "Organization ID"
// @Success 200 {object} response.SuccessResponse{data=dto.OrganizationResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Router /organizations/{id} [get]
func GetOrganization(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid organization ID"))
	}

	org, err := coreOrg.GetByID(c.Context(), id)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, org)
}

// CreateOrganization handles POST /organizations
// @Summary Create organization
// @Description Create a new organization
// @Tags organizations
// @Accept json
// @Produce json
// @Param request body dto.CreateOrganizationRequest true "Organization details"
// @Success 201 {object} response.SuccessResponse{data=dto.OrganizationResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 401 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /organizations [post]
func CreateOrganization(c *fiber.Ctx) error {
	var req coreOrg.CreateRequest
	if err := c.BodyParser(&req); err != nil {
		return response.Error(c, core.InvalidInputError("Invalid request body"))
	}
	if err := validation.ValidateStruct(req); err != nil {
		return response.Error(c, err)
	}

	org, err := coreOrg.Create(c.Context(), req)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Created(c, org)
}

// UpdateOrganization handles PUT /organizations/:id
// @Summary Update organization
// @Description Update an existing organization
// @Tags organizations
// @Accept json
// @Produce json
// @Param id path string true "Organization ID"
// @Param request body dto.UpdateOrganizationRequest true "Organization update details"
// @Success 200 {object} response.SuccessResponse{data=dto.OrganizationResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /organizations/{id} [put]
func UpdateOrganization(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid organization ID"))
	}

	var req coreOrg.UpdateRequest
	if err := c.BodyParser(&req); err != nil {
		return response.Error(c, core.InvalidInputError("Invalid request body"))
	}

	org, err := coreOrg.Update(c.Context(), id, req)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, org)
}

// DeleteOrganization handles DELETE /organizations/:id
// @Summary Delete organization
// @Description Delete an organization by ID
// @Tags organizations
// @Accept json
// @Produce json
// @Param id path string true "Organization ID"
// @Success 200 {object} response.SuccessResponse{data=object{success=boolean}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /organizations/{id} [delete]
func DeleteOrganization(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid organization ID"))
	}

	if err := coreOrg.Delete(c.Context(), id); err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, fiber.Map{"success": true})
}

// JoinOrganization handles POST /organizations/:id/join
// @Summary Join organization
// @Description Join an organization as the current user
// @Tags organizations
// @Accept json
// @Produce json
// @Param id path string true "Organization ID"
// @Success 200 {object} response.SuccessResponse{data=object{success=boolean}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 401 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /organizations/{id}/join [post]
func JoinOrganization(c *fiber.Ctx) error {
	userID, err := middleware.GetUserID(c)
	if err != nil {
		return response.Error(c, core.UnauthorizedError("Not authenticated"))
	}

	idStr := c.Params("id")
	orgID, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid organization ID"))
	}

	if err := coreOrg.JoinOrganization(c.Context(), userID, orgID); err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, fiber.Map{"success": true})
}

// LeaveOrganization handles POST /organizations/leave
// @Summary Leave organization
// @Description Leave the current organization
// @Tags organizations
// @Accept json
// @Produce json
// @Success 200 {object} response.SuccessResponse{data=object{success=boolean}}
// @Failure 401 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /organizations/leave [post]
func LeaveOrganization(c *fiber.Ctx) error {
	userID, err := middleware.GetUserID(c)
	if err != nil {
		return response.Error(c, core.UnauthorizedError("Not authenticated"))
	}

	if err := coreOrg.LeaveOrganization(c.Context(), userID); err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, fiber.Map{"success": true})
}

File: backend/src/api/organization/handlers.rs
use axum::{
    Json,
    extract::{Path, State, rejection::JsonRejection},
    http::StatusCode,
};
use uuid::Uuid;

use crate::api::response::SuccessResponse;

use super::{
    dto::{
        OrganizationCreateRequest, OrganizationResponse, OrganizationUpdateRequest, SuccessFlag,
    },
    error::{OrganizationApiError, OrganizationAuthenticated},
    state::OrganizationState,
};

type ApiResult<T> = Result<Json<SuccessResponse<T>>, OrganizationApiError>;

fn organization_id(value: &str) -> Result<Uuid, OrganizationApiError> {
    Uuid::parse_str(value).map_err(|_| {
        crate::error::AppError::InvalidInput("Invalid organization ID".to_owned()).into()
    })
}

#[utoipa::path(
    get,
    path = "/organizations",
    responses(
        (status = 200, body = SuccessResponse<Vec<OrganizationResponse>>),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "organizations",
    operation_id = "listOrganizations"
)]
pub async fn list_organizations(
    _authenticated: OrganizationAuthenticated,
    State(state): State<OrganizationState>,
) -> ApiResult<Vec<OrganizationResponse>> {
    Ok(Json(SuccessResponse::new(
        state
            .service()
            .list()
            .await?
            .into_iter()
            .map(Into::into)
            .collect(),
    )))
}

#[utoipa::path(
    post,
    path = "/organizations",
    request_body = OrganizationCreateRequest,
    responses(
        (status = 201, body = SuccessResponse<OrganizationResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 409, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "organizations",
    operation_id = "createOrganization"
)]
pub async fn create_organization(
    _authenticated: OrganizationAuthenticated,
    State(state): State<OrganizationState>,
    body: Result<Json<OrganizationCreateRequest>, JsonRejection>,
) -> Result<(StatusCode, Json<SuccessResponse<OrganizationResponse>>), OrganizationApiError> {
    let Json(request) = body.map_err(|_| OrganizationApiError::invalid_body())?;
    let organization = state.service().create(request.validate()?).await?;
    Ok((
        StatusCode::CREATED,
        Json(SuccessResponse::new(organization.into())),
    ))
}

#[utoipa::path(
    get,
    path = "/organizations/{id}",
    params(("id" = Uuid, Path)),
    responses(
        (status = 200, body = SuccessResponse<OrganizationResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "organizations",
    operation_id = "getOrganization"
)]
pub async fn get_organization(
    _authenticated: OrganizationAuthenticated,
    State(state): State<OrganizationState>,
    Path(id): Path<String>,
) -> ApiResult<OrganizationResponse> {
    Ok(Json(SuccessResponse::new(
        state
            .service()
            .get_by_id(organization_id(&id)?)
            .await?
            .into(),
    )))
}

#[utoipa::path(
    put,
    path = "/organizations/{id}",
    params(("id" = Uuid, Path)),
    request_body = OrganizationUpdateRequest,
    responses(
        (status = 200, body = SuccessResponse<OrganizationResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 409, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "organizations",
    operation_id = "updateOrganization"
)]
pub async fn update_organization(
    _authenticated: OrganizationAuthenticated,
    State(state): State<OrganizationState>,
    Path(id): Path<String>,
    body: Result<Json<OrganizationUpdateRequest>, JsonRejection>,
) -> ApiResult<OrganizationResponse> {
    let id = organization_id(&id)?;
    let Json(request) = body.map_err(|_| OrganizationApiError::invalid_body())?;
    Ok(Json(SuccessResponse::new(
        state.service().update(id, request.into()).await?.into(),
    )))
}

#[utoipa::path(
    delete,
    path = "/organizations/{id}",
    params(("id" = Uuid, Path)),
    responses(
        (status = 200, body = SuccessResponse<SuccessFlag>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "organizations",
    operation_id = "deleteOrganization"
)]
pub async fn delete_organization(
    _authenticated: OrganizationAuthenticated,
    State(state): State<OrganizationState>,
    Path(id): Path<String>,
) -> ApiResult<SuccessFlag> {
    state.service().delete(organization_id(&id)?).await?;
    Ok(Json(SuccessResponse::new(SuccessFlag { success: true })))
}

#[utoipa::path(
    post,
    path = "/organizations/{id}/join",
    params(("id" = Uuid, Path)),
    responses(
        (status = 200, body = SuccessResponse<SuccessFlag>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "organizations",
    operation_id = "joinOrganization"
)]
pub async fn join_organization(
    OrganizationAuthenticated(account_id): OrganizationAuthenticated,
    State(state): State<OrganizationState>,
    Path(id): Path<String>,
) -> ApiResult<SuccessFlag> {
    state
        .service()
        .join_organization(account_id.into_inner(), organization_id(&id)?)
        .await?;
    Ok(Json(SuccessResponse::new(SuccessFlag { success: true })))
}

#[utoipa::path(
    post,
    path = "/organizations/leave",
    responses(
        (status = 200, body = SuccessResponse<SuccessFlag>),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "organizations",
    operation_id = "leaveOrganization"
)]
pub async fn leave_organization(
    OrganizationAuthenticated(account_id): OrganizationAuthenticated,
    State(state): State<OrganizationState>,
) -> ApiResult<SuccessFlag> {
    state
        .service()
        .leave_organization(account_id.into_inner())
        .await?;
    Ok(Json(SuccessResponse::new(SuccessFlag { success: true })))
}
